src/compiler/checker.ts

```diff
@@ -44446,77 +44452,125 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
-                const right = evaluate((expr as BinaryExpression).right, location);
+            }
+            case SyntaxKind.BinaryExpression: {
+                const evaluatedLeft = evaluate((expr as BinaryExpression).left, location);
+                if (!evaluatedLeft) {
+                    break;
+                }
+                const evaluatedRight = evaluate((expr as BinaryExpression).right, location);
+                if (!evaluatedRight) {
+                    break;
+                }
+                const left = evaluatedLeft.value;
+                const right = evaluatedRight.value;
                 if (typeof left === "number" && typeof right === "number") {
                     switch ((expr as BinaryExpression).operatorToken.kind) {
                         case SyntaxKind.BarToken:
-                            return left | right;
+                            return { value: left | right };
                         case SyntaxKind.AmpersandToken:
-                            return left & right;
+                            return { value: left & right };
                         case SyntaxKind.GreaterThanGreaterThanToken:
-                            return left >> right;
+                            return { value: left >> right };
                         case SyntaxKind.GreaterThanGreaterThanGreaterThanToken:
-                            return left >>> right;
+                            return { value: left >>> right };
                         case SyntaxKind.LessThanLessThanToken:
-                            return left << right;
+                            return { value: left << right };
                         case SyntaxKind.CaretToken:
-                            return left ^ right;
+                            return { value: left ^ right };
                         case SyntaxKind.AsteriskToken:
-                            return left * right;
+                            return { value: left * right };
                         case SyntaxKind.SlashToken:
-                            return left / right;
+                            return { value: left / right };
                         case SyntaxKind.PlusToken:
-                            return left + right;
+                            return { value: left + right };
                         case SyntaxKind.MinusToken:
-                            return left - right;
+                            return { value: left - right };
                         case SyntaxKind.PercentToken:
-                            return left % right;
+                            return { value: left % right };
                         case SyntaxKind.AsteriskAsteriskToken:
-                            return left ** right;
+                            return { value: left ** right };
+                    }
+                }
+                else if (typeof left === "bigint" && typeof right === "bigint") {

```

Comment:
This is somewhat redundant since it doesn't really close "gap" between the template literal types simplification and this evaluation here. I feel like it would be weird for those 2 to behave differently though:
```ts
enum T {
  A = `${1 +1 }`,
  B = `${1n +1n }`
}
```

On top of that, it mostly repeats the number-oriented case but I don't know how to implement this in a better way since I can't force TS to treat two different identifiers to be of the same type. It always complains about `number | bigint` in this scenario (for good reasons).